futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+    /// already been called, the future will complete immediately without making
+    /// any further progress.
+    pub fn new(future: T, reg: AbortRegistration) -> Self {
+        Abortable {
+            future: future,
+            inner: reg.inner,
+        }
+    }
+
+    unsafe_pinned!(future: T);
+}
+
+/// A registration handle for a `Abortable` future.
+/// Values of this type can be acquired from `AbortHandle::new` and are used
+/// in calls to `Abortable::new`.
+#[derive(Debug, Clone)]
+pub struct AbortRegistration {
+    inner: Arc<AbortInner>,
+}
+
+/// A handle to a `Abortable` future.
+#[derive(Debug, Clone)]
+pub struct AbortHandle {
+    inner: Arc<AbortInner>,
+}
+
+impl AbortHandle {
+    /// Creates a new `AbortHandle` which can be used to abort a running future.
+    /// This function also returns a `AbortRegistration` which is used to
+    /// pair the handle with a `Abortable` future.
+    ///
+    /// This function is usually paired with a call to `Abortable::new`.
+    pub fn new() -> (Self, AbortRegistration) {
+        let inner = Arc::new(AbortInner {
+            waker: AtomicWaker::new(),
+            cancel: AtomicBool::new(false),
+        });
+
+        (
+            AbortHandle {
+                inner: inner.clone(),
+            },
+            AbortRegistration {
+                inner: inner,
+            },
+        )
+    }
+}
+
+// Inner type storing the waker to awaken and a bool indicating that it
+// should be cancelled.
+#[derive(Debug)]
+struct AbortInner {
+    waker: AtomicWaker,
+    cancel: AtomicBool,
+}
+
+/// Creates a new `Abortable` future and a `AbortHandle` which can be used to stop it.
+///
+/// This function is a convenient (but less flexible) alternative to calling
+/// `AbortHandle::new` and `Abortable::new` manually.
+pub fn abortable<T>(future: T) -> (Abortable<T>, AbortHandle)
+    where T: Future
+{
+    let (handle, reg) = AbortHandle::new();
+    (
+        Abortable::new(future, reg),
+        handle,
+    )
+}
+
+/// Indicator that the `Abortable` future was aborted.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Aborted;
+
+impl<T> Future for Abortable<T> where T: Future {
+    type Output = Result<T::Output, Aborted>;
+
+    fn poll(mut self: PinMut<Self>, cx: &mut task::Context) -> Poll<Self::Output> {
+        // Check if the future has been aborted
+        if self.inner.cancel.load(Ordering::Relaxed) {
+            return Poll::Ready(Err(Aborted))
+        }
+
+        // attempt to complete the future
+        if let Poll::Ready(x) = self.future().poll(cx) {
+            return Poll::Ready(Ok(x))
+        }
+
+        // Register to receive a wakeup if the future is aborted in the... future
+        self.inner.waker.register(cx.waker());
+
+        // Check to see if the future was aborted between the first check and
+        // registration
+        if self.inner.cancel.load(Ordering::Relaxed) {
+            return Poll::Ready(Err(Aborted))
+        }
```

Comment:
I don't think this will work. I think `Ordering::Relaxed` makes this check unreliable.

My suggestion for `poll`:
```rust
// Register to receive a wakeup if the future is aborted in the... future
self.inner.waker.register(cx.waker());

// Check if the future has been aborted
if self.inner.cancel.load(Ordering::Aquire) {
    return Poll::Ready(Err(Aborted))
}

// Attempt to complete the future
match self.future().poll(cx) {
    Poll::Ready(output) => Poll::Ready(Ok(output)),
    Poll::Pending => Poll::Pending,
}
```

And `abort`:
```
self.inner.cancel.store(true, Ordering::Release);
self.inner.waker.wake();
```

I might be wrong of course, because atomic ops are not exactly my forte

Reply:
The call to `register` introduces an `AcqRel` memory barrier.

Reply:
- The API docs for `AtomicWaker::wake` should mention this
- A comment about this that explains why `Ordering::Relaxed` is enough would be good
- I still think that one check of `self.inner.cancel` per `poll` call should suffice

Reply:
> The API docs for AtomicWaker::wake should mention this

There's [an open issue about documenting this](https://github.com/rust-lang-nursery/futures-rs/issues/942), and it has yet to be decided exactly what guarantees we want to offer.

>  A comment about this that explains why Ordering::Relaxed is enough would be good

Done

> I still think that one check of self.inner.cancel per poll call should suffice

It does not-- either we need to register to receive a wakeup unconditionally (prior to checking the value of `cancel` or polling the future) or we need to check both before and after.